# Oh My Py — канал про тайные возможности стандартной библиотеки Питона. 
- [ ]  https://t.me/ohmypy/4

Тайные не потому, что кто-то их скрывает, конечно ツ

Просто стандартная библиотека огромная! А разработчики часто не копают глубоко и изобретают велосипед вместо того, чтобы использовать готовое.

---

## Сделать превьюшку длинного текста

Допустим, мы хотим получить превьюшку длинной статьи. Можно обрезать механически:



In [1]:
article = "Около двух месяцев назад породистый голубь по имени Френк постучался в стеклянные двери омской ветеринарной клиники"
article[:30]

'Около двух месяцев назад пород'

Фраза оборвана посреди слова — это неуважение к читателю и к Френку.

А можно воспользоваться функцией `textwrap.shorten()`:


In [2]:
import textwrap

textwrap.shorten(article, 30, placeholder="...")


'Около двух месяцев назад...'

## Отформатировать текст для консоли

Если любите делать CLI-утилиты, модуль textwrap наверняка вам понравится.

Он умеет перформатировать многострочный текст, чтобы длина строки не превышала N символов:


In [3]:
text = "Около двух месяцев назад породистый голубь по имени Френк постучался в стеклянные двери омской ветеринарной клиники"
formatted = textwrap.fill(text, width=20)
print(formatted)


Около двух месяцев
назад породистый
голубь по имени
Френк постучался в
стеклянные двери
омской ветеринарной
клиники


In [4]:
# Или добавить отступ, например для цитаты:

import textwrap
inspirational = "Цитаты простых людей:"
quote = "Откройте окно вообще дышать невозможно"
quote = textwrap.indent(quote, prefix="> ")
print(inspirational, quote, sep="\n")


Цитаты простых людей:
> Откройте окно вообще дышать невозможно


## Все слова с прописной буквы

Допустим, запустили вы стартап. В автоматическом режиме собираете самые упоротые новости русскоязычных СМИ, вот такие:

_Кот из Новокузнецка признан виновным в потопе_

Автоматически же переводите их на английский, вот так:

_Cat from Novokuznetsk found guilty in the flood_

И ежедневно рассылаете подписчикам по всему миру.

Всё хорошо, но знакомый эксперт из МГИМО подсказывает: в английском принято каждое слово в заголовке начинать с заглавной буквы. А у вас-то не так!

Можно, конечно, бить заголовок по пробелам 
- через `.split()`, исправлять регистр 
- через `.capitalize()` и склеивать обратно 
- через `.join()`. 

Но есть способ лучше:


In [5]:
import string
header = "Cat from Novokuznetsk found guilty in the flood"
string.capwords(header)

'Cat From Novokuznetsk Found Guilty In The Flood'

## Простое сравнение с шаблоном

Для проверки строки по шаблону обычно используют регулярные выражения и модуль re. Но иногда хочется что-нибудь попроще, пусть и не такое мощное — вроде `like` в SQL.

Сравнить строку или список с шаблоном поможет модуль `fnmatch`:


In [6]:
import fnmatch
journal = [
  "10:00 Начался обычный день в омской ветклинике",
  "10:30 Голубь Френк постучался в стеклянные двери",
  "10:50 Лисица Клер поскреблась в окно",
  "11:10 Попугай Питер проник через вентиляцию",
  "11:11 Клер попыталась сожрать Френка и Питера",
  "11:25 Осьминог Пауль всплыл в мужском туалете",
]

fnmatch.filter(journal, "*Френк*")


['10:30 Голубь Френк постучался в стеклянные двери',
 '11:11 Клер попыталась сожрать Френка и Питера']

In [7]:
fnmatch.fnmatch("frank", "f???k")

True

> Под капотом используются регулярки, так что всегда можно конвертировать шаблон в регулярное выражение:

In [8]:
fnmatch.translate("*Френк*")

'(?s:.*\\Ф\\р\\е\\н\\к.*)\\Z'

## Сравнить строки на похожесть

Помните ваш стартап с самыми актуальными новостями дня? Кажется, у него появился конкурент — он нагло крадёт ваши аутентичные новости, рерайтит их, и рассылает ничего не подозревающим клиентам, подрывая вашу репутацию.

Судите сами, вот ваши новости:

In [1]:
genuine = [
  "«Братец-хлеб» из Китая носит плащ и корону из булочек, чтобы кормить чаек",
  "Мясо гигантских тараканов станет вкусной и недорогой альтернативой говядине",
  "Скандал в ботаническом саду: 10 миллионов рублей ушло на зарплату кактусам",
]

А вот новости жалкого подражателя:


In [2]:
plagiary = [
  "Китайский хлебный братец кормит чаек плащом и короной из булочек",
  "Гигантское мясо тараканов станет говядине недорогой и вкусной альтернативой",
  "Зарплата кактусов в ботаническом саду составила 10 скандальных миллионов рублей",
]

Нужны какие-то основания для судебного иска, и нужны быстро. Хорошо, что в стандартной библиотеке Питона есть модуль `difflib`. Сделаем на нём функцию сравнения:


In [3]:
import difflib

def similarity(s1, s2):
  normalized1 = s1.lower()
  normalized2 = s2.lower()
  matcher = difflib.SequenceMatcher(None, normalized1, normalized2)
  return matcher.ratio()


In [4]:
similarity(genuine[0], plagiary[0])

0.5109489051094891

In [6]:
similarity(genuine[1], plagiary[1])

0.6933333333333334

In [8]:
similarity(genuine[2], plagiary[2])

0.5490196078431373

### _АГА!!!_ 

— $51\%$, $69\%$ и $55\%$ похожести!

_Всё ясно_, какие ещё нужны доказательства.